tests/screenshots/build.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+use std::io::Write;
+use std::path::Path;
+
+/// Returns a list of all the `.slint` files in the `tests/cases` subfolders.
+pub fn collect_test_cases() -> std::io::Result<Vec<test_driver_lib::TestCase>> {
+    let mut results = vec![];
+
+    let case_root_dir: std::path::PathBuf = [env!("CARGO_MANIFEST_DIR"), "cases"].iter().collect();
+
+    println!("cargo:rerun-if-env-changed=SLINT_TEST_FILTER");
+    let filter = std::env::var("SLINT_TEST_FILTER").ok();
+
+    for entry in walkdir::WalkDir::new(case_root_dir.clone()).follow_links(true) {
+        let entry = entry?;
+        let absolute_path = entry.into_path();
+        if absolute_path.is_dir() {
+            println!("cargo:rerun-if-changed={}", absolute_path.display());
+            continue;
+        }
+        let relative_path =
+            std::path::PathBuf::from(absolute_path.strip_prefix(&case_root_dir).unwrap());
+        if let Some(filter) = &filter {
+            if !relative_path.to_str().unwrap().contains(filter) {
+                continue;
+            }
+        }
+        if let Some(ext) = absolute_path.extension() {
+            if ext == "60" || ext == "slint" {
+                results.push(test_driver_lib::TestCase { absolute_path, relative_path });
+            }
+        }
+    }
+    Ok(results)
+}
+
+fn main() -> std::io::Result<()> {
+    let mut generated_file = std::fs::File::create(
+        Path::new(&std::env::var_os("OUT_DIR").unwrap()).join("generated.rs"),
+    )?;
+
+    let references_root_dir: std::path::PathBuf =
+        [env!("CARGO_MANIFEST_DIR"), "references"].iter().collect();
+
+    for (i, testcase) in collect_test_cases()?.into_iter().enumerate() {
+        let case_root_dir: std::path::PathBuf =
+            [env!("CARGO_MANIFEST_DIR"), "cases"].iter().collect();
+
+        let reference_path = testcase.absolute_path.strip_prefix(case_root_dir).unwrap();
+        let mut reference_path =
+            references_root_dir.join(reference_path).with_extension("png").display().to_string();
+
+        reference_path = format!("\"{}\"", reference_path);
+
+        println!("cargo:rerun-if-changed={}", testcase.absolute_path.display());
+        let mut module_name = testcase.identifier();
+        if module_name.starts_with(|c: char| !c.is_ascii_alphabetic()) {
+            module_name.insert(0, '_');
+        }
+        writeln!(generated_file, "#[path=\"{0}.rs\"] mod r#{0};", module_name)?;
+        let source = std::fs::read_to_string(&testcase.absolute_path)?;
+
+        let mut output = std::fs::File::create(
+            Path::new(&std::env::var_os("OUT_DIR").unwrap()).join(format!("{}.rs", module_name)),
+        )?;
+
+        generate_source(source.as_str(), &mut output, testcase).unwrap();
+
+        let create_screenshots = std::env::var("SLINT_CREATE_SCREENSHOTS").ok();
```

Comment:
Personally, i would do this check at runtime rather then at compile time.

If you want to do it at compile time, you need to print the right cargo invocation so the build script is re-run if that env variable change.

(The adventage of doing the test at runtime is that you don't need to rebuild the whole testcase, which can be slow, to take the screenshot in case of faillure)

Reply:
That flag is now only intended to create a example references.

Reply:
I know,  but the workflow would be to run the test,  see the faillure, then run the test again with this flag. Right?

right now, this workflow don't work because nothing would be rebuild by changing just the flag.

Reply:
This flag is not part of the workflow. This is only a helper to create initial screenshots to compare to (the stick of truth). The stuff we think how it should looks like.

Reply:
What i mean is that setting SLINT_CREATE_SCREENSHOTS will not rebuild anything and therefore no screenshot will be re-created.  (unless you do another change somewhere)